lib/using_restricted_array.rb

```diff
@@ -1,82 +1,140 @@
-require_relative 'restricted_array.rb'
+require_relative "restricted_array.rb"
+
 # RestrictedArray can be created using a specified size, or a random size in
 # the range of 1-20 will be chosen for you.
 # All values are integers in the range of 1-221.
 # RestrictedArray cannot be resized.
 
 # Calculates the length of the restricted array. All values are integers.
 # The restricted_array is terminated by 'nil' i.e. array[length] = nil
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity:linear or O(n), n is the value_to_find of items in the array or the length of the input array
+# Space complexity: O(1)
 def length(array)
-  raise NotImplementedError
+  i = 0
+  until array[i] == nil
+    i += 1
+  end
+  return i
 end
 
 # Prints each integer values in the array
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: linear or O(n)
+# Space complexity: O(1)
 def print_array(array)
-  raise NotImplementedError
+  len = length(array)
+  i = 0
+  while i < len
+    print array[i]
+    print " "
+    i += 1
+  end
+  puts
 end
 
 # For an unsorted array, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: linear or O(n)
+# Space complexity: O(1)
 def search(array, length, value_to_find)
-  raise NotImplementedError
+  length.times do |i|
+    if array[i] == value_to_find
+      return true
+    end
+  end
+  return false
 end
 
 # Finds and returns the largest integer value the array
 # Assumes that the array is not sorted.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: linear or O(n)
+# Space complexity: O(1)
 def find_largest(array, length)
-  raise NotImplementedError
+  largest_number = array[0]
+  length.times do |i|
+    if array[i] > largest_number
+      largest_number = array[i]
+    end
+  end
+  return largest_number
 end
 
 # Finds and returns the smallest integer value in the array
 # Assumes that the array is not sorted.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: linear or O(n)
+# Space complexity: O(1)
 def find_smallest(array, length)
-  raise NotImplementedError
+  smallest_number = array[0]
+  length.times do |i|
+    if smallest_number > array[i]
+      smallest_number = array[i]
+    end
+  end
+  return smallest_number
 end
 
 # Reverses the values in the integer array in place
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(1)
 def reverse(array, length)
-  raise NotImplementedError
+  n = 1
+  (length / 2).times do |i|
+    temp = 0
+    temp = array[i]
+    array[i] = array[length - n]
+    array[length - n] = temp
```

Comment:
What you have is 💯, but in case you haven't seen it yet, there's another way to swap in Ruby:

```
array[i], array[length - n] = array[length - n], array[i]
```